common/script.c

```diff
@@ -0,0 +1,426 @@
+/*
+ * Phoenix-RTOS
+ *
+ * psu script parsing
+ *
+ * Copyright 2020 Phoenix Systems
+ *
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include "script.h"
+
+#define LOG_ERROR(...)  fprintf(stderr, __VA_ARGS__)
+#define LOG(...)        fprintf(stdout, __VA_ARGS__)
+
+#define IS_SPACE(c)     ((c) == ' '  || (c) == '\t')
+#define IS_NEWLINE(c)   ((c) == '\n' || (c) == '\r')
+#define IS_DIGIT(c)     ((c) >= '0'  && (c) <= '9' )
+#define IS_ALPHA_LOW(c) ((c) >= 'a'  && (c) <= 'z' )
+#define IS_ALPHA_UP(c)  ((c) >= 'A'  && (c) <= 'Z' )
+#define IS_ALPHA(c)     (IS_ALPHA_LOW(c) || IS_ALPHA_UP(c))
+#define IS_QUOTE(c)     ((c) == '\"' || (c) == '\'')
+
+
+/* initialize parser, and load psu script file */
+int script_load(script_t *s, const char *fname)
+{
+	int fd;
+	void *ptr;
+	struct stat statbuf = { 0 };
+
+	if ((fd = open(fname, O_RDONLY)) < 0) {
+		LOG_ERROR("Unable to open '%s' script file.\n", fname);
+
+		return SCRIPT_ERROR;
+	}
+
+	if (fstat(fd, &statbuf) < 0) {
+		LOG_ERROR("fstat() failed.\n");
+		close(fd);
+
+		return SCRIPT_ERROR;
+	}
+
+	if (statbuf.st_size > 1024*100) {
+		LOG_ERROR("Script file too big (100kB limit).\n");
+		close(fd);
+
+		return SCRIPT_ERROR;
+	}
+
+	ptr = mmap(NULL, statbuf.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
+
+	close(fd);
+
+	if (ptr == MAP_FAILED) {
+		LOG_ERROR("Unable to allocate memory.\n");
+
+		return SCRIPT_ERROR;
+	}
+
+	memset(s, 0, sizeof(*s));
+	s->ptr = ptr;
+	s->buf.ptr = ptr;
+	s->buf.end = ptr + statbuf.st_size;
+
+	return SCRIPT_OK;
+}
+
+
+/* free parser context */
+void script_close(script_t *s)
+{
+	if (s->buf.ptr == NULL || s->buf.ptr == MAP_FAILED) {
+		return;
+	}
```

Comment:
According to our coding convention, one-liners don't need brackets. Only exceptions is case, when one or more else-if body(s) have more than one line. Eg:

```
if (cond)
	do_something();
```

```
if (cond) {
	do_something();
}
else {
	do_something_else();
	do_the_other_thing();
}
```

Same rule applies to loops, but not to functions ;)